pandas/core/indexes/multi.py

```diff
@@ -906,7 +906,12 @@ def _get_level_values(self, level):
         labels = self.labels[level]
         filled = algos.take_1d(unique._values, labels,
                                fill_value=unique._na_value)
-        values = unique._shallow_copy(filled)
+
+        if not unique._can_hold_na and filled.dtype == 'float':
```

Comment:
i would always use the infer method no need to have an if

Reply:
This breaks several tests with dates (e.g. in ``pandas/tests/test_multilevel.py`` and ``pandas/tests/indexes/test_multi.py``), in one case because _"ValueError: Inferred frequency None from passed dates does not conform to passed frequency D"_, in other cases because (I think) ``DateTimeIndex``es with time zone are _re_added the time zone (if I understand correctly).

Reply:
you need to see if you can find a general soln. we don't want to have if/then cases all over the place.

Reply:
The only "general" solution I can think is a property e.g. ``_na_upcast`` of ``Index`` subclasses (inherited from ``Index``) which is the subclass itself if ``_can_hold_na = True`` (that is, all but ``Int64Index``), and is overridden to ``Float64Index`` for ``Int64Index``. Then I would just call ``unique._na_upcast._shallow_copy(filled)``.

Do you have in mind anything better?

Reply:
i have no idea ; you need to see why other tests are failing and work around